### ISSUE TITLE
Hide platform boilerplate code in Flutter projects

### ISSUE THREAD

**Issue by roman-petrov** [flutter/flutter#59043]
## Use case

Currently each multi-platform Flutter project contains significant amount of platform boilerplate code. For example, if I have enabled just `Android`, `iOS`, `Web` and `Windows` platforms and run `flutter create myproject` I get `~250K` boilerplate code in `android`, `ios`, `web` and `windows` folders.

All this code generally should be checked in into VCS.

Problems with this approach:
- It is quite hard to modify some platform-specific project settings for each platform. For example, to change project display name user should look into all these folders containing quite complex code especially for non-experts.
- Currently it seems Flutter does not have a good workflow scenario to update this boilerplate code when project templates are updated/improved in Flutter. The only approach I see is to run `flutter create` again and overwrite all this code with new templates. But any changes made here might be lost.
- It is almost impossible to create a new Flutter project "by hand" without ``flutter create`.
- The amount of boilerplate code might be much bigger than application code. A `Hello, world` application source code base might become even close to `1 MB` in future if targeted to all supported platforms (Android, iOS, MacOS, Web, Windows, Linux, Fuchsia, etc). 

All these problems might become more important with growing number of supported platforms in Flutter.

## Proposal

The proposal is to hide all platform boilerplate code / project templates in Flutter build tools and provide some cross-platform settings file(s) (maybe just use `pubspec.yaml`?). So a simple application might contain only `main.dart` and `pubspec.yaml` file and problems above will be solved.

This is just a draft proposal and I understand that serious amount of work should be done in Flutter tooling and to achieve this. I love Flutter and just hope that this feature might make Flutter better and simplify Flutter learning curve, especially for beginners.

**Comment by stuartmorgan** [comment]
@Hixie This sounds a lot like your ephemeral vs. materialized runner thoughts; are those tracked/discussed somewhere already?

**Comment by Hixie** [comment]
> It is quite hard to modify some platform-specific project settings for each platform. For example, to change project display name user should look into all these folders containing quite complex code especially for non-experts.

This wouldn't be solved by hiding the code except for the features we happen to implement.


> Currently it seems Flutter does not have a good workflow scenario to update this boilerplate code when project templates are updated/improved in Flutter. The only approach I see is to run flutter create again and overwrite all this code with new templates. But any changes made here might be lost.

This won't really be solved either by hiding the code by default, it would only solve the problem for people who don't modify the code. Lots of applications require this code to be modified though, for various platform-specific features to be accessed.


> It is almost impossible to create a new Flutter project "by hand" without flutter create.

It would still be impossible. I mean, it would be equivalent to today creating a pubspec.yaml and a lib/main.dart and then running `flutter create .`.


> The amount of boilerplate code might be much bigger than application code. A Hello, world application source code base might become even close to 1 MB in future if targeted to all supported platforms (Android, iOS, MacOS, Web, Windows, Linux, Fuchsia, etc).

This will still be true, it just wouldn't be checked into the repo.

That said, I entirely support this proposal, because I think that even though this isn't actually solving all these problems, it does make it better for new users. It reduces the length of the on-ramp before you get to something useful. It moves the problems with updates and so forth to past the point where you care about those files, which is good because if you don't care about them then you shouldn't need to worry about them.

> This sounds a lot like your ephemeral vs. materialized runner thoughts; are those tracked/discussed somewhere already?

I think there's bugs filed on these ideas already, but this bug is a fine description of the problem also.

**Comment by roman-petrov** [comment]
@Hixie, thank you for your comments. I understand that there is no silver bullet to solve all the problems I listed above.

I also understand that users currently often *have to* modify platform-specific code especially in complex applications. The first thing I suggest here is to provide a simple workflow for *simple* cases and retain current (*more complex*) workflow for all other cases.

In my proposal I forgot to list another (possible complicated use case) which (*not easily*) might be solved in Flutter tooling.

## Use case

Consider user needs to add platform-specific plugin to his project. For many plugins he just needs to add it to `pubspec.yaml`. But some plugins require manual changes in platform-specific code in corresponding directories, for example, [firebase](https://pub.dev/packages/firebase).

When I did this first time, I was unable to find an detailed official guide on a single page about how to make these changes. I found some Medium articles first, then found some official instructions at

<https://firebase.google.com/docs/flutter/setup?platform=android>
<https://firebase.google.com/docs/flutter/setup?platform=ios>
<https://pub.dev/packages/firebase_web#using-this-package-for-browser-applications>

So I finally spent a couple of hours or even more when integrating `Firebase` to my Android/Web applications first time. 

# Proposal

Actually, I am trying to "continue" the proposal I filled in this issue. The idea is to allow plugins to "contribute" to platform specific code in Flutter build process. 

For Android, plugins may contribute gradle plugins to be injected into `android/build.gradle` / `android/app/build.gradle` , for Web - JavaScript files to be injected into `index.html` and so on.

I understand that this proposal might be a major update to Flutter build system / plugin tooling, but I think this one might be included into "phase 2" if/after the initial proposal is implemented somehow.

**Comment by pedromassango** [comment]
> For Android, plugins may contribute gradle plugins to be injected into `android/build.gradle` / `android/app/build.gradle` , for Web - JavaScript files to be injected into `index.html` and so on.

I agree with you on this feature. but now what about plugins that requires a minimum SDK version and the one provide in the plugin is different of the current Flutter min SDK version?

**Comment by roman-petrov** [comment]
Just putting some related issues here:

- [Enable the Flutter tool to change app id after project creation](https://github.com/flutter/flutter/issues/13212)
- [Add a flutter project validation mechanism](https://github.com/flutter/flutter/issues/12983)
- [Add a project upgrade mechanism](https://github.com/flutter/flutter/issues/12573)
- [Please copy the description in the generated web/manifest.json from pubspec.yaml](https://github.com/flutter/flutter/issues/48742)
- [Document why files are in the template](https://github.com/flutter/flutter/issues/9884)

**Comment by roman-petrov** [comment]
A related Flutter design document: [Automatically update project files after changing the version of the Flutter SDK](https://docs.google.com/document/d/1fPnpUI6gXE9PkNWuFoAW20m_sQ87ZSC1rrn_D7fxBvs)

**Comment by roman-petrov** [comment]
Just found similar idea in a proposal coming from Flutter team (only for web projects, solves just one case) https://github.com/flutter/flutter/pull/80519#issuecomment-825099567

**Comment by roman-petrov** [comment]
There is a community package [change_app_package_name](https://pub.dev/packages/change_app_package_name) which attempts to solve one of issues with currently used approach.

**Comment by roman-petrov** [comment]
https://github.com/flutter/flutter/issues/78254

**Comment by roman-petrov** [comment]
https://github.com/nhancv/nft

**Comment by roman-petrov** [comment]
Related PR https://github.com/flutter/flutter/pull/95708
